### PR TITLE
Beats event processing and default fields

### PR DESCRIPTION
**Pull request by urso**
This changes moves the generation of the event processing into it's
distinct package, such that the actual publisher pipeline will not
define any processors anymore. A new instance of a publisher pipeline
must not add fields on it's own.

With this change we convert the event processing pipline into the 'Supporter'
pattern, which is already used for Index Management.
As different beats ask for slightly different behavior in the event
processing (e.g. normalize, default builtins and so on), the
`processing.Supporter` can be used for customizations.

Also fixes new fields accidentily being added to the monitoring outputs, as it separates the pipeline and processors.

Simplifies tests, but also adds a few test cases for dynamic fields and other settings.

### CODE REVIEW

File: libbeat/publisher/processing/processing.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// the publisher pipeline.  The factory gets the global configuration passed,
+// in order to configure some shared global event processing.
+type SupporterFactory func(info beat.Info, log *logp.Logger, cfg *common.Config) (Supporter, error)
+
+// Supporter is used to create an event processing pipeline. It is used by the
+// publisher pipeline when a client connects to the pipeline. The supporter
+// will merge the global and local configurations into a common event
+// processor.
+// If `drop` is set, then the processor generated must always drop all events.
+type Supporter func(cfg beat.ProcessingConfig, drop bool) (beat.Processor, error)
```
Comment (kvch) [inline libbeat/publisher/processing/processing.go]: To me this name is too abstract for what the function does, like calling a type `Handler`. What do you think about naming these `processing.ConfigApplier` and `processing.ConfigApplierFactory` or `processing.ConfigResolver` and `processing.ConfigResolverFactory`. So the concrete functions would be named `processing.NewBeatConfigApplier`/`processing.NewBeatConfigResolver` and `processing.NewObserverConfigApplier`/`processing.NewObserverConfigResolver`. WDYT?

Comment (urso) [reply libbeat/publisher/processing/processing.go]: The names mostly mimic index management which says `idxmgmt.Supporter` and so on. I used the same name for having a consistent naming scheme on similar functionality/patterns. Names `ilm.Supporter`, `idxmgmt.Supporter`, and `processing.Supporter`. Common patterns should have somewhat common names.
Maybe a better named would have been `Provider` or `Feature`, as I see these  as overwritable libbeat features that originally have been hard coded.

While the constructors get an input config, I'd not name them `ConfigApplier` or `ConfigResolver`. If possible use a single noun over a multili-noun name. E.g. `Configurer` or `Featurer`. (I try not to make my names sound like twitter messages)

All these packages somewhat follow the strategy, abstract factory, and factory method patterns. The later mostly by chance.

The pattern found in ilm, idxmgmt, and now processing goes like this:
- beat instance use `SupporterFactory` in order to provide it's own subsystems with some actual features implementation (Strategy):
  - The `SupportFactory` is feed with the beats global configuration in order to unpack configs and prepare its own state.
- Supporter is the actual feature/strategy presented to a sub-system.
- ilm, idxmgmt, publisher pipeline Subsystems act themselves as builders/constructors/factories for other subsystems that can pass additional parameters.
  - examples:
    - Elasticsearch output uses idxmgmt subsystem on connect
    - ES output uses idxmgmt subsystem for creating an index selector
    - setup uses idxmgmt subsystem with custom ES output
    - publisher pipeline uses processing in order to setup the final event processing pipeline.
  - As these subsystems act as Factories for other components the `ilm/idxmgmt/processing.Supporter` follows the `Factory Method` pattern. e.g. `BuildSelector`, `Manager` . This is no commonalty, but only by chance, as the other subsystems expect Factories.
- The final instances generated by the `Supporter(s)` are the Strategies that run within a subsystems context: e.g. idxmgmt.Manager, beat.Processor.

Actually we don't really need the definitions of Supporter/SupporterFactory in these packages. Alternative we can remove them and move the interface definitions to the instance package. But as they are still passed around into other packages I defined these interface types for convenience.

I'm fine with finding better, names in general. How about `Feature` and `FeatureFactory`? No matter which names we choose, we will have to update the other packages in a followup PR as well.

Note: The `processing.Supporter` should be an interface, no function type.

Comment (kvch) [reply libbeat/publisher/processing/processing.go]: Thank you for the detailed explanation. In this case I am ok with sticking with `Supporter`. I haven't been able to come up with anything better. :(
+1 on being consistent over packages.

File: libbeat/publisher/processing/default.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+	"fmt"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+	"github.com/elastic/beats/libbeat/processors/actions"
+)
+
+type pipelineProcessors struct {
```
Comment (simitt) [inline libbeat/publisher/processing/default.go]: Why is this called `pipelineProcessors` but part of the `publisher/processing` package? 

I am generally having a bit of a hard time to follow along with the terminology (processing, processor, processors, pipeline and pipelineProcessor).

File: libbeat/cmd/instance/beat.go [hunk 3]
```diff
@@ -593,6 +597,16 @@ func (b *Beat) configure(settings Settings) error {
 		imFactory = idxmgmt.MakeDefaultSupport(settings.ILM)
 	}
 	b.index, err = imFactory(nil, b.Beat.Info, b.RawConfig)
+	if err != nil {
+		return err
+	}
+
+	processingFactory := settings.Processing
```
Comment (simitt) [inline libbeat/cmd/instance/beat.go]: Why is the `ProcessorFactory` defined on the root cmd level, but the `ProcessingConfig` on the pipeline level? Would it make sense to move the processorFactory also to the processing config level?

**Comment by ph** [comment]
OKm I went through the code here, LGTM, I also agree with @simitt's comment that if we can align the naming that would be great even if the packages have different responsabilities.

File: libbeat/publisher/processing/default.go [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+	fields      common.MapStr
+	tags        []string
+
+	processors beat.Processor
+
+	drop       bool // disabled is set if outputs have been disabled via CLI
+	alwaysCopy bool
+}
+
+const ecsVersion = "1.0.0-beta2"
```
[the remark this answers is not in the document]
Comment (ph) [reply libbeat/publisher/processing/default.go]: The above comment is more, how we will keep the value in sync between release, I think we will forget. if we rely on the version from the official package the version will be updated every time we update the vendor libraries.

**Comment by ph** [comment]
> I think in the long run it would make sense to also export the single processors, so they are reusable. At the moment, if one needs a small adaption, the whole Processor implementation would need to be re-implemented.

I agree with the above.

File: libbeat/publisher/pipeline/pipeline.go [hunk 5]
```diff
@@ -347,7 +317,16 @@ func (p *Pipeline) ConnectWith(cfg beat.ClientConfig) (beat.Client, error) {
...
+	if p.processors != nil {
+		proc, err := p.processors.Create(cfg.Processing, publishDisabled)
+		if err != nil {
+			return nil, err
+		}
+
+		processors = proc
+	}
+
 	acker := p.makeACKer(processors != nil, &cfg, waitClose)
```
[the remark this answers is not in the document]
Comment (urso) [reply libbeat/publisher/pipeline/pipeline.go]: processors.Create could still return `nil`.

File: libbeat/publisher/processing/processors.go [hunk 6]
```diff
@@ -27,7 +27,6 @@ import (
 	"github.com/elastic/beats/libbeat/logp"
 	"github.com/elastic/beats/libbeat/outputs/codec/json"
 	"github.com/elastic/beats/libbeat/processors"
-	"github.com/elastic/beats/libbeat/processors/actions"
 )
 
 type program struct {
```
[the remark this answers is not in the document]
Comment (urso) [reply libbeat/publisher/processing/processors.go]: renaming to 'group'.

File: libbeat/publisher/processing/default.go (hunk 2 above)
Comment (urso) [reply libbeat/publisher/processing/default.go]: Most code is moved as is. Renamed to 'builder'.

File: libbeat/cmd/instance/beat.go (hunk 3 above)
Comment (urso) [reply libbeat/cmd/instance/beat.go]: Beats do not directly publish methods to the queue in the publisher pipeline, but do so via a `Client`.
The pipeline including global processors and fields settings, queue, and outputs is created on startup. There are no go-routines collecting data yet.
Go-routines are supposed to use `Connect`, so to connect to a publisher pipeline. This returns a `beat.Client` (Client instances are not guaranteed to be thread safe). Each client/go-routine is allowed to configure local processors/fields/tags, which gets merged with the global settings. The factory is the global entity, loading, checking, and preparing global configurations on startup. The ProcessingConfig specifies the per go-routine local processing, which might be eventually established due to the Beat modules initialization or much later via autodiscovery.

**Comment by urso** [comment]
I refrained from exporting more functionality for now. The PR is already big enough. There is quite some legacy code regarding processors all over the place, that needs cleanup. I actually extracted the `processors.AddFields/AddTags` processors from this original code base a few weeks ago. If there is need we can export some more. The only noteworthy processors we might export in the future are: `generalizeProcessor`, and `debugPrintProcessor`.
Maybe the `group` one, but this one overlaps with some other legacy functionality in the processors package :(

Most of the logic is actually within the builder itself.

File: libbeat/publisher/processing/default.go [hunk 7]
```diff
@@ -20,71 +20,75 @@ package processing
...
-	// of a pipieline client.
+	// of a pipeline client.
 	ClientFields(beat.Info, beat.ProcessingConfig) common.MapStr
 }
 
 type builtinModifier func(beat.Info) common.MapStr
 
-// NewBeatSupport creates a new SupporterFactory based on NewDefaultSupport.
-// NewBeatSupport automatically adds the `ecs.version`, `host.name` and `agent.X` fields
+// MakeDefaultBeatSupport creates a new SupporterFactory based on NewDefaultSupport.
```
Comment (simitt) [inline libbeat/publisher/processing/default.go]: nit: comment not updated to `SupportFactory`

File: libbeat/publisher/processing/default_test.go [hunk 8]
```diff
@@ -163,7 +164,7 @@ func TestProcessorsConfigs(t *testing.T) {
 		"with client processor": {
 			local: beat.ProcessingConfig{
 				Processor: func() beat.ProcessorList {
-					p := newProgram("test", logp.L())
+					p := newGroup("test", logp.L())
```
Comment (simitt) [inline libbeat/publisher/processing/default_test.go]: nit: also rename the variable `p`.

File: libbeat/publisher/processing/default.go (hunk 7 above)
Comment (urso) [reply libbeat/publisher/processing/default.go]: ah, gorename fun :)

File: libbeat/publisher/processing/processors.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+	if p == nil || len(p.list) == 0 {
+		return event, nil
+	}
+
+	for _, sub := range p.list {
+		var err error
+
+		event, err = sub.Run(event)
+		if err != nil {
+			// XXX: We don't drop the event, but continue filtering here iff the most
```
Comment (ph) [inline libbeat/publisher/processing/processors.go]: nit: iff -> if

**Comment by urso** [comment]
Travis was green, but codecov upload failed.
Unrelated metricbeat test failed with connection problem.

**Pull request merged**